### PR TITLE
SCM improvements

### LINKED ISSUE #3039: [SCM] Recipe with sources in the same repo copies to the source folder gitignored files
Situation:
I have a `conanfile.py` in a repo with:
```
scm = {
        "type": "git",
        "subfolder": "asdio",
        "url": "auto",
        "revision": "auto"
    }
```
There are some gitignored files such as `.vs`.

Problem:
When using `conan install -pr=...` it uploads *everything* including ignored files to the `source` folder.

### LINKED ISSUE #3044: [SCM] Cloning repository not working when using exports attribute
- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

## System info
**Conan version**: 1.4.4
**Operating System**: Debian 9 (stretch)

## Problem explanation
When using scm and also specify at least a file in the "exports" attribute, conan fails to bring the sources with git clone, saying that the folder is not empty. Thing is that before doing git clone in the source folder in local cache, conan first copies what is in the "export" folder to the "source" folder. Obviously, you can't do git clone afterwards.
If specifying a subfolder, scm would work, but I do need the sources to be in the root "source" folder, not in a subfolder.

## Steps to reproduce
I reproduced it in a repo: https://github.com/Sebiee/conan_bug
```bash
git clone https://github.com/Sebiee/conan_bug
cd conan_bug
conan export . conan/test
#remove file scm_folder.txt from local cache to force cloning from git
conan test test_package string_wrapper/1.0.0@conan/test --build string_wrapper
```

### LINKED ISSUE #3069: [SCM] Using subfolder with a repository together with the conanfile
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Working with Conan 1.4.4 (also tested with 1.4.3) on OSX, we expect to rely on the (experimental) [SCM attribute of `ConanFile`](https://docs.conan.io/en/latest/creating_packages/package_repo.html#capturing-the-remote-and-commit-from-git-scm-experimental) to implement CI of our repositories (#3003 for context)

Here is a minimal conanfile:
```
from conans import ConanFile, CMake, tools

class ScmtestConan(ConanFile):
    name = "scm-test"
    version = "0.0.0"
    description = "Try the SCM conanfile attribute"
    settings = "os", "compiler", "build_type", "arch"

    generators = "cmake", "cmake_paths"

    scm = {
        "type": "git",
        "subfolder": "cloned_repo",
        "url": "auto",
        "revision": "auto"
    }

    def build(self):
        cmake = CMake(self)
        cmake.configure(source_folder="cloned_repo")
        cmake.build()
        cmake.install()

    def package(self):
        # Done by the install step in build()
        pass

    def package_info(self):
        self.cpp_info.libs = tools.collect_libs(self)
```

Now, if we export and try to install the recipe with `conan export . com/develop && conan install scm-test/0.0.0@com/develop`, the build fails.

### subfolder problem

After inspecting the created directories under `~/.conan`, it turn out that, even though a `cloned_repo` is created in the `source` folder, the repo is actually cloned directly into the `source` folder, and `cloned_repo` remains empty.

### same computer optimization problem

Now, if we were to move the conanfile.py from the root of the repo to a `conan/` subfolder, as we intend to, nothing is cloned at all, because it seems (following the TIP) that conan wants to use the local folder when installing on the same computer that did the export, and it uses the folder containing the conanfile, not necessarily the folder at the root of the repo.

### PR DESCRIPTION
**Pull request by lasote**
- [x] Refer to the issue that supports this Pull Request. Closes #3039  Closes #3069  Closes #3044

### CODE REVIEW

File: conans/client/source.py [hunk 1]
```diff
@@ -42,15 +34,35 @@ def complete_recipe_sources(remote_manager, client_cache, registry, conanfile, c
...
+    for src_dir, dirs, files in os.walk(src, followlinks=True):
+        def is_excluded(the_dir):
+            if the_dir in excluded:
+                return True
+            # If dst is a sub-folder of src_dir avoid infinite loop
+            if cmn_prefix == src:  # dst is a subfolder
+                rel_dst = os.path.relpath(dst, src)
+                return rel_dst.split(os.path.sep)[0] == the_dir
+        # Overwriting the dirs will prevents walk to get into them
+        if src_dir == src:  # First step, root folder
```
Comment (memsharded) [inline conans/client/source.py]: Why only the root are excluded and not children?

File: conans/client/cmd/export.py [hunk 2]
```diff
@@ -104,24 +104,29 @@ def _capture_export_scm_data(conanfile, src_path, destination_folder, output, pa
     if os.path.exists(scm_src_file):
         os.unlink(scm_src_file)
 
-    scm = get_scm(conanfile, src_path)
-
-    if not scm or not (scm.capture_origin or scm.capture_revision):
+    try:
```
[the remark this answers is not in the document]
Comment (lasote) [reply conans/client/cmd/export.py]: I thought just the opposite, much more explicit, almost the same lines, only twice in the code.

**Comment by lasote** [comment]
I'm reviewing the `merge_directories` function and adding some more test.

**Comment by lasote** [comment]
Review again.

**Pull request merged**